pkg/radar/center.go

```diff
@@ -0,0 +1,37 @@
+package radar
+
+import (
+	"github.com/dharmab/skyeye/pkg/coalitions"
+	"github.com/dharmab/skyeye/pkg/encyclopedia"
+	"github.com/paulmach/orb"
+	"github.com/paulmach/orb/geo"
+	"github.com/rs/zerolog/log"
+)
+
+func shiftPoint(a orb.Point, b orb.Point) orb.Point {
+	if a.Lon() == 0 && a.Lat() == 0 {
+		return b
+	}
+	return geo.Midpoint(a, b)
+}
+
+func (s *scope) updateCenterPoint() {
+	blue := orb.Point{}
+	red := orb.Point{}
+	itr := s.contacts.itr()
+	for itr.next() {
```

Comment:
I dont know if there's a go reason for this, but wouldn't this be more readable? I guess maybe you have some custom iterator functionality? I suppose you can't implement a custom range in go.

```go
for _, contact := range s.contacts {
     ...
```

Reply:
Using `range` here alone is unsafe because the collection of contacts is being updated by other goroutines concurrently. One could use a `sync.RWMutex` to obtain a read lock while ranging, but you also need to remember to lock and unlock every time the collection is accessed. `radar.contactsDatabase` wraps the collection with locking in a way that's easier to use and test, at the tradeoff of not exposing direct access to the collection. One could also use a `sync.Map`, but the contactsDatabase also provides some indexing functionality so it's still useful.

What you're seeing here is a weakness of Go compared to Java/C#; the collection types you get in the stdlib are very barebones. (Go doesn't even have a HashSet in the stdlib!). The expectation is that higher level data structures are either implemented by the programmer specific to the use case (as here with the contacts database) or you use a third-party module (such as the deque backing each trackfile).

Reply:
Heh, [sync,Map](https://pkg.go.dev/sync#Map) actually says not to use it for this case:

> Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content. 

In this case we have two indexes within the contacts database that need to be invariant.

Reply:
Python handles this super nicely as well, they let you [define your own iterators that do work in the way you'd expect](https://wiki.python.org/moin/Iterator), and they have a [nice toolkit in the stdlib](https://docs.python.org/3/library/itertools.html) for common patterns.